pulse2percept/utils.py

```diff
@@ -1,603 +1,607 @@
+    kernel : array_like
+        Second input, typically the kernel
+    mode : str {'full', 'valid', 'same'}, optional
+        A string indicating the size of the output:
+        ``full``
+        The output is the full discrete linear convolution of the inputs.
+        (Default)
+        ``valid``
+        The output consists only of those elements that do not rely on
+        zero-padding.
+        ``same``
+        The output is the same size as `data`, centered with respect to the
+        'full' output.
+    method : str {'fft', 'sparse'}, optional
+        A string indicating the convolution method:
+        ``fft``
+        Use the fast Fourier transform (FFT). (Default)
+        ``sparse``
+        Use the sparse convolution.
+    dojit : bool, optional
+        A flag indicating whether to use numba's just-in-time compilation
+        option (only relevant for `method`=='sparse').
+    """
+    if method.lower() == 'fft':
+        # Use FFT: faster on non-sparse data
+        conved = tsample * sps.fftconvolve(data, kernel, mode)
+    elif method.lower() == 'sparse':
+        # Use sparseconv: faster on sparse data
+        conved = tsample * sparseconv(data, kernel, mode, dojit)
+    else:
+        raise ValueError("Acceptable methods are: 'fft', 'sparse'.")
+    return conved
+
+
+def center_vector(vec, newlen):
+    """
+    Returns the center `newlen` portion of a vector.
+
+    Adapted from scipy.signal.signaltools._centered:
+    github.com/scipy/scipy/blob/v0.18.0/scipy/signal/signaltools.py#L236-L243
+
+    """
+    currlen = vec.shape[-1]
+    startind = (currlen - newlen) // 2
+    endind = startind + newlen
+    return vec[startind:endind]
+
+
+def parfor(func, in_list, out_shape=None, n_jobs=-1, engine='joblib',
+           backend='threading', func_args=[], func_kwargs={}):
+    """
+    Parallel for loop for numpy arrays
+
+    Parameters
+    ----------
+    func : callable
+        The function to apply to each item in the array. Must have the form:
+        func(arr, idx, *args, *kwargs) where arr is an ndarray and idx is an
+        index into that array (a tuple). The Return of `func` needs to be one
+        item (e.g. float, int) per input item.
+
+    in_list : list
+       All legitimate inputs to the function to operate over.
+
+    n_jobs : integer, optional
+        The number of jobs to perform in parallel. -1 to use all cpus
+        Default: 1
+
+    engine : str
+        {"dask", "joblib", "serial"}
+        The last one is useful for debugging -- runs the code without any
+        parallelization.
+
+    backend : str
+        What dask backend to use. Irrelevant for other engines.
+
+    func_args : list, optional
+        Positional arguments to `func`
+
+    func_kwargs : list, optional
+        Keyword arguments to `func`
+
+    Returns
+    -------
+    ndarray of identical shape to `arr`
+
+    Notes
+    -----
+    Imported from pyAFQ (blob e20eaa0 from June 3, 2016):
```

Comment:
At some point I noticed that pyAFQ had an updated `parfor` with respect to ours, so I imported the new version and "tagged" it accordingly.